### PR TITLE
Turn on single-file analyzer and fix problems

### PR DESCRIPTION
**Pull request by agocke**
Should fully fix https://github.com/dotnet/runtime/issues/44488

## Proposed changes

This change turns on the single-file analyzer in the winforms repo, fixes the few real problems that were encountered, and suppresses warnings when issues are addressed.

## Customer Impact

Full support for single-file publishing in WinForms, with some confidence that all issues are addressed.

## Regression? 

- No

## Risk

Very low.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/5426)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlVersionInfo.cs [hunk 1]
```diff
@@ -161,6 +165,7 @@ internal string ProductVersion
             ///  Retrieves the FileVersionInfo associated with the main module for
             ///  the component.
             /// </summary>
+            [RequiresAssemblyFiles("Throws if _owner is an in-memory assembly. Check " + nameof(OwnerIsInMemoryAssembly) + " first")]
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlVersionInfo.cs:168]:
```suggestion
            [RequiresAssemblyFiles("Throws if " + nameof(_owner) + " is an in-memory assembly. Check " + nameof(OwnerIsInMemoryAssembly) + " first")]
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlVersionInfo.cs [hunk 2]
```diff
@@ -172,6 +177,9 @@ private FileVersionInfo GetFileVersionInfo()
 
                 return _versionInfo;
             }
+
+            //[UnconditionalSuppressMessage("SingleFile", "IL3000")]
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlVersionInfo.cs:181]: Is this commented out code needed?
```suggestion
            // [UnconditionalSuppressMessage("SingleFile", "IL3000")]
```

**Review by RussKie — DISMISSED** [review]
👍

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlVersionInfo.cs (hunk 2 above)
Comment (agocke) [reply src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlVersionInfo.cs:181]: Woops, good catch! That shouldn't be necessary now.

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs [hunk 3]
```diff
@@ -951,14 +950,15 @@ internal static void FormActivated(bool modal, bool activated)
...
+        [UnconditionalSuppressMessage("SingleFile", "IL3002", Justification = "Single-file case is handled")]
         private static FileVersionInfo GetAppFileVersionInfo()
         {
             lock (s_internalSyncObject)
             {
                 if (s_appFileVersion is null)
                 {
                     Type t = GetAppMainType();
-                    if (t is not null)
+                    if (t is not null && t.Assembly.Location.Length > 0)
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/Application.cs:961]: What made to add this second condition here?

Comment (agocke) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs:961]: If this is in a single-file, `t.Module.FullyQualifiedName` will return `<Unknown>`  and `GetVersionInfo` will fail. `ExecutablePath` should be single-file safe, however.

**Review by dreddy-work — APPROVED** [review]

**Pull request merged**